# 📊 Exploratory Data Analysis (EDA)

## School Bullying Analysis using the CRDC 2021–22 Dataset

**Notebook:** 03 – Exploratory Data Analysis

**Author:** Yoldas Erdem

**Project:** Data Analytics & AI Bootcamp Capstone

---

# 📖 Introduction

Following the completion of the ETL pipeline, this notebook marks the beginning of the analytical phase of the project.

The cleaned CRDC Harassment and Bullying dataset is explored to better understand its structure, assess its overall quality, and identify meaningful patterns before conducting SQL analysis and building interactive Tableau dashboards.

Rather than testing predefined hypotheses, the purpose of this notebook is to become familiar with the data, uncover potential relationships, and identify questions worth investigating in greater depth.

Exploratory Data Analysis (EDA) serves as a critical bridge between data preparation and statistical analysis. The findings documented throughout this notebook will directly inform the SQL queries, dashboard design, and final presentation.

---

# 🎯 Objectives

This notebook aims to:

- Understand the overall structure of the cleaned dataset.
- Verify that the ETL process produced a reliable analytical dataset.
- Explore the distribution of bullying allegations across different categories.
- Investigate demographic patterns among affected students.
- Examine disciplinary actions reported by schools.
- Identify geographic trends across states and school districts.
- Generate analytical questions for the SQL phase.
- Define potential visualizations for the Tableau dashboard.

---

# 🗺️ Notebook Roadmap

The exploratory analysis is organized into three major sections.

## Section 1 — Dataset Understanding

Gain an understanding of the dataset before beginning any analysis.

Topics include:

- Dataset dimensions
- Data types
- Missing values
- Duplicate records
- Metadata overview
- Summary statistics

---

## Section 2 — Data Exploration

Explore the main components of the CRDC dataset and identify meaningful patterns.

Topics include:

- School metadata
- Bullying allegations
- Reported students
- Disciplinary actions
- Geographic comparisons
- Distribution analysis
- Correlation analysis

---

## Section 3 — Key Findings

Summarize the most important observations and prepare for the analytical phase of the project.

Topics include:

- Key insights
- Potential limitations
- Questions for SQL analysis
- Ideas for Tableau dashboards

---

# 🔄 Analytics Workflow

This notebook represents the third stage of the overall project workflow.

```

Raw CRDC Dataset
        │
        ▼
ETL Pipeline
        │
        ▼
Clean Analytical Dataset
        │
        ▼
📍 Exploratory Data Analysis
        │
        ▼
PostgreSQL Database
        │
        ▼
SQL Analysis
        │
        ▼
Tableau Dashboard
        │
        ▼
Executive Presentation

```

---

# 🚀 Let's Begin

The analysis starts with a high-level assessment of the cleaned dataset to ensure its structure, completeness, and quality before exploring individual variables in greater detail.

# 1. Dataset Understanding

## Purpose

Before exploring individual variables, it is important to understand the overall structure and quality of the dataset.

This section provides a high-level overview of the cleaned CRDC dataset by examining its dimensions, data types, completeness, and basic descriptive statistics. These initial checks confirm that the ETL process produced a reliable analytical dataset and establish a foundation for the exploratory analyses that follow.

---

### Analysis Checklist

- Load the cleaned dataset
- Review dataset dimensions
- Inspect data types
- Examine missing values
- Verify duplicate records
- Review summary statistics
- Inspect the first observations

# 1.1 Load the Clean Dataset

## Purpose

The exploratory analysis begins by loading the cleaned dataset produced during the ETL pipeline.

Unlike the original CRDC dataset, this version contains standardized variable names, preserved identifier fields, and CRDC reserve codes converted to missing values (`NULL`). It serves as the primary analytical dataset used throughout the remainder of this project.

After loading the dataset, a preview of the first observations is displayed to verify that the data was imported successfully.

In [19]:
# =====================================================
# Import Required Libraries
# =====================================================

from pathlib import Path
import pandas as pd

In [20]:
# =====================================================
# Project Configuration
# =====================================================

# Define the project root directory
project_root = Path.cwd().parent

# Path to the cleaned dataset created during the ETL pipeline
clean_data = project_root / "data" / "processed" / "crdc_bullying_clean.csv"

In [21]:
# =====================================================
# Load the Clean Dataset
# =====================================================

df = pd.read_csv(
    clean_data,
    dtype={
        "district_id": str,
        "school_key": str
    },
    low_memory=False
)

# Display the first five records
df.head()

,state_code,state,district_id,district,school_id,school,school_key,juvenile_justice_school,allegation_sex,allegation_orientation,...,disciplined_disability_multi_race_male,disciplined_disability_multi_race_female,total_disciplined_disability_male,total_disciplined_disability_female,disciplined_disability_english_learner_male,disciplined_disability_english_learner_female,disciplined_disability_idea_male,disciplined_disability_idea_female,disciplined_disability_section_504_male,disciplined_disability_section_504_female
0,AL,ALABAMA,0100002,Alabama Youth Services,99995,AUTAUGA CAMPUS,010000299995,Yes,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,AL,ALABAMA,0100005,Albertville City,870,Albertville Middle School,010000500870,No,2.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,AL,ALABAMA,0100005,Albertville City,871,Albertville High School,010000500871,No,4.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,AL,ALABAMA,0100005,Albertville City,879,Albertville Intermediate School,010000500879,No,1.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,AL,ALABAMA,0100005,Albertville City,889,Albertville Elementary School,010000500889,No,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


# 1.2 Dataset Overview

## Purpose

Before examining individual variables, it is essential to understand the overall structure of the dataset.

This section provides a high-level overview, including the number of observations, variables, data types, and memory usage. These statistics help verify that the dataset has been loaded correctly and provide context for the analyses that follow.

In [22]:
# =====================================================
# Dataset Summary
# =====================================================

dataset_summary = pd.DataFrame({
    "Metric": [
        "Schools (Rows)",
        "Variables (Columns)"
    ],
    "Value": [
        f"{df.shape[0]:,}",
        df.shape[1]
    ]
})

dataset_summary

,Metric,Value
0,Schools (Rows),"98,010"
1,Variables (Columns),159


In [23]:
# =====================================================
# Dataset Information
# =====================================================

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 98010 entries, 0 to 98009
Columns: 159 entries, state_code to disciplined_disability_section_504_female
dtypes: float64(151), int64(1), str(7)
memory usage: 118.9 MB


In [24]:
# =====================================================
# Summary Statistics
# =====================================================

summary_statistics = df.describe().T

summary_statistics.head(15)

,count,mean,std,min,25%,50%,75%,max
school_id,98010.0,3926.152944,9765.355942,0.0,753.0,1785.0,4184.75,99999.0
allegation_sex,91902.0,0.641738,5.135084,0.0,0.0,0.0,0.00,1133.0
allegation_orientation,91902.0,0.251453,4.564510,0.0,0.0,0.0,0.00,1133.0
allegation_race,91902.0,0.456922,4.917722,0.0,0.0,0.0,0.00,1133.0
allegation_disability,91902.0,0.172564,4.562140,0.0,0.0,0.0,0.00,1133.0
allegation_religion,91902.0,0.029346,0.367293,0.0,0.0,0.0,0.00,43.0
allegation_atag,91777.0,0.000338,0.027017,0.0,0.0,0.0,0.00,5.0
allegation_budd,91777.0,0.000022,0.004668,0.0,0.0,0.0,0.00,1.0
allegation_cath,91777.0,0.000131,0.013203,0.0,0.0,0.0,0.00,2.0
allegation_east,91777.0,0.000044,0.006602,0.0,0.0,0.0,0.00,1.0


# 1.3 Variable Overview

## Purpose

Understanding the variables available in the dataset is an important first step in exploratory analysis.

This section categorizes the variables according to their analytical purpose. Organizing the dataset into logical groups provides a clearer understanding of its structure and establishes the foundation for the analyses performed throughout the remainder of this notebook.

The CRDC Harassment and Bullying dataset can be divided into four main categories:

- **Metadata** – School and district identifiers.
- **Allegations** – Reported harassment and bullying allegations.
- **Reported Students** – Students reported as affected by bullying.
- **Disciplined Students** – Students receiving disciplinary actions following reported incidents.

In [25]:
# =====================================================
# Categorize Variables
# =====================================================

# Metadata variables
metadata_cols = [
    "state_code",
    "state",
    "district_id",
    "district",
    "school_id",
    "school",
    "school_key",
    "juvenile_justice_school"
]

# Allegation variables
allegation_cols = [
    col for col in df.columns
    if col.startswith("allegation_")
]

# Reported student variables
reported_cols = [
    col for col in df.columns
    if col.startswith("reported_") or col.startswith("total_reported_")
]

# Disciplined student variables
disciplined_cols = [
    col for col in df.columns
    if col.startswith("disciplined_") or col.startswith("total_disciplined_")
]

In [26]:
# =====================================================
# Variable Category Summary
# =====================================================

variable_summary = pd.DataFrame({
    "Variable Category": [
        "Metadata",
        "Bullying Allegations",
        "Reported Students",
        "Disciplined Students"
    ],
    "Number of Variables": [
        len(metadata_cols),
        len(allegation_cols),
        len(reported_cols),
        len(disciplined_cols)
    ]
})

variable_summary.loc[len(variable_summary)] = [
    "Total",
    variable_summary["Number of Variables"].sum()
]

variable_summary

,Variable Category,Number of Variables
0,Metadata,8
1,Bullying Allegations,19
2,Reported Students,66
3,Disciplined Students,66
4,Total,159


### Interpretation

The cleaned dataset contains **159 analytical variables**, organized into four logical categories.

The majority of variables describe either **students reported as affected by bullying (66 variables)** or **students receiving disciplinary actions (66 variables)**. These two groups account for more than 80% of all variables and represent the primary focus of the analytical phase.

A smaller group of **19 allegation variables** records the different types of harassment and bullying reported by schools, while **8 metadata variables** provide the geographic and institutional context required for comparisons across schools, districts, and states.

This categorization provides a clear roadmap for the exploratory analysis that follows.

# 1.4 Missing Values

## Purpose

Missing values are an expected characteristic of the cleaned CRDC dataset.

During the ETL process, official CRDC reserve codes (e.g., `-3`, `-4`, `-5`, `-6`, `-9`, `-12`, and `-13`) were intentionally converted to missing values (`NULL`) to improve data quality and prevent these administrative codes from being interpreted as legitimate numerical values.

This section evaluates the overall completeness of the dataset and identifies where missing values remain.

In [27]:
# =====================================================
# Missing Value Summary
# =====================================================

missing_summary = pd.DataFrame({
    "Metric": [
        "Total Missing Values",
        "Columns with Missing Values",
        "Complete Columns"
    ],
    "Value": [
        df.isna().sum().sum(),
        (df.isna().sum() > 0).sum(),
        (df.isna().sum() == 0).sum()
    ]
})

missing_summary

,Metric,Value
0,Total Missing Values,446839
1,Columns with Missing Values,151
2,Complete Columns,8


In [28]:
# =====================================================
# Top Variables with Missing Values
# =====================================================

missing_columns = (
    df.isna()
      .sum()
      .sort_values(ascending=False)
      .reset_index()
)

missing_columns.columns = [
    "Variable",
    "Missing Values"
]

missing_columns.head(15)

,Variable,Missing Values
0,allegation_hind,6405
1,allegation_prot,6233
2,allegation_atag,6233
3,allegation_sikh,6233
4,allegation_othrel,6233
5,allegation_othchrn,6233
6,allegation_multi,6233
7,allegation_morm,6233
8,allegation_jwsh,6233
9,allegation_jwit,6233


In [29]:
# =====================================================
# Missing Value Percentage
# =====================================================

missing_columns["Missing (%)"] = (
    missing_columns["Missing Values"]
    / len(df)
    * 100
).round(2)

missing_columns.head(15)

,Variable,Missing Values,Missing (%)
0,allegation_hind,6405,6.54
1,allegation_prot,6233,6.36
2,allegation_atag,6233,6.36
3,allegation_sikh,6233,6.36
4,allegation_othrel,6233,6.36
5,allegation_othchrn,6233,6.36
6,allegation_multi,6233,6.36
7,allegation_morm,6233,6.36
8,allegation_jwsh,6233,6.36
9,allegation_jwit,6233,6.36


### Interpretation

The cleaned dataset contains **446,839 missing values** across **151 variables**, while the **8 metadata columns are complete**. This is expected because the ETL pipeline converted official CRDC reserve codes into `NULL` values.

The missing values are concentrated mainly in the analytical measurement columns, especially the religion-related allegation variables. For example, `allegation_hind` has the highest missing count with **6,405 missing values**, representing **6.54%** of all rows. Most other religion subcategory variables have around **6.36%** missing values.

This means the dataset remains highly usable: even the variables with the most missing values are still available for more than **93%** of schools.

The missing values should not be treated as errors or deleted automatically. They represent administrative reporting cases such as unavailable, not applicable, or suppressed data. Therefore, they will be retained as `NULL` values and handled carefully during SQL analysis and Tableau visualization.

# 1.5 Duplicate Records

## Purpose

Before conducting any analysis, it is important to verify that each observation represents a unique school.

Duplicate records could distort summary statistics, aggregate calculations, and visualizations by counting the same school more than once. This section confirms whether duplicate observations are present in the cleaned dataset.

In [30]:
# =====================================================
# Duplicate Record Check
# =====================================================

duplicate_count = df.duplicated().sum()

duplicate_summary = pd.DataFrame({
    "Metric": ["Duplicate Records"],
    "Value": [duplicate_count]
})

duplicate_summary

,Metric,Value
0,Duplicate Records,0


### Interpretation

No duplicate records were identified in the cleaned dataset.

This confirms that each row represents a unique school and that the ETL pipeline preserved the original school-level structure without introducing duplicate observations. The dataset is therefore suitable for aggregation and comparative analysis.

# 1.6 Metadata Summary

## Purpose

The metadata variables provide important contextual information about the schools included in the dataset.

This section summarizes the overall coverage of the CRDC dataset by reporting the number of schools, states, school districts, and juvenile justice schools represented in the cleaned dataset.

In [31]:
# =====================================================
# Metadata Summary
# =====================================================

metadata_summary = pd.DataFrame({
    "Metric": [
        "States",
        "School Districts",
        "Schools",
        "Juvenile Justice Schools"
    ],
    "Value": [
        df["state"].nunique(),
        df["district_id"].nunique(),
        df["school_id"].nunique(),
        (df["juvenile_justice_school"] == "Yes").sum()
    ]
})

metadata_summary

,Metric,Value
0,States,52
1,School Districts,17704
2,Schools,13907
3,Juvenile Justice Schools,882


### Interpretation

The CRDC Harassment and Bullying dataset provides nationwide coverage of public schools across the United States.

The dataset contains **98,010 school records**, representing schools from every U.S. state and the District of Columbia. In addition to traditional public schools, the dataset also includes juvenile justice schools, allowing bullying and harassment patterns to be explored across different educational settings.

The complete metadata enables reliable comparisons between schools, districts, and states throughout the remainder of the project.

# 1.7 Dataset Understanding Summary

## Key Findings

The initial assessment confirms that the cleaned CRDC dataset is suitable for exploratory analysis.

Key observations include:

- The dataset contains **98,010 school records** described by **159 analytical variables**.
- Variables are organized into four logical categories: metadata, bullying allegations, reported students, and disciplined students.
- All metadata fields are complete, providing reliable geographic and institutional identifiers.
- Missing values are expected and result from official CRDC reserve codes converted to `NULL` values during the ETL process.
- No duplicate records were identified.

Overall, the dataset demonstrates a high level of completeness and provides a reliable foundation for the exploratory analyses presented in the following section.